### DISCUSSION TITLE
A few questions

### DISCUSSION THREAD

**Discussion by loofou** [exodrifter/unity-rumor#75] [General]
Hey exodrifter,

I started to implement Rumor into my game, but I'm running into a few problems that are probably related to me not fully understanding the capabilities of the expression system yet.

First of all: Do you have support for comments? Basically: Can I comment out lines of rumor code for it to be ignored fully by the parser? I tried to look into the parser code, but it's a bit more 

Second question: The current error messages are often not very helpful. I guess that comes down to how the Parser Combinator is set up. Do you have some tips on how to interpret the error messages? How do you usually go about this and figure out what's wrong? For example: the index the error message is specifying is difficult to interpret, as I'm not sure if it strips any whitespace for this or not.

Third question: Can it be that Functions are not fully supported for Bindings yet? I can link, bind and use Actions no problem, but functions fail already on the compile step. Right now I'm using a work-around by setting a "return_value" variable on the scope from an action call, but it's very hacky.

I hope you can help with this,
Loofou

**Comment by exodrifter** [comment]
1. Yes. It appears I failed to mention it in the documentation, but any line starting with `#` should be ignored by the compiler.
2. Indeed, the error messages are not very useful right now. If the list of cases is particularly long, it's usually because of a missing character on a previous line or improper indentation. The compiler doesn't strip any whitespace so the position information should be accurate, but even so the position information isn't super helpful because of the fact that generally speaking the compiler might fail to parse the current line (where it will report the error) due to some actual error present on the previous line(s). If you are willing to post an example and the error you got from it, I can probably help you figure out what's wrong.
3. ~~Functions should be fully supported just like actions, though perhaps there's a bug I am unaware of. What kind of compilation error do you get?~~ I looked into the code briefly and it seems like while I added the types to support functions, I forgot to add it to the parser. I will have to fix that next week.

**Reply by loofou** [reply]
1. Ah, I will test the hash again. I thought I did already, but maybe there was another error present somewhere.
Edit: Tested it and no. any # at the beginning of a line results in a parser error.

2. It was more a general question. I sometimes get random errors and not really sure in which line they are, etc. What would help is if the index would get converted into a "line, character in line" output. This way it would be much easier to find the location of the parsing error. It seems you already have a place to aggregate the exceptions the parser throws, so there might be a good point to clean up the messages as well, maybe. But I understand that that's quite an effort.
3. Yeah it felt like that. Generally it seemed it was implemented everywhere, but the parser, but I wasn't sure if you just re-used the parser for the actions instead.

Thanks!

**Reply by exodrifter** [reply]
1. Uh... I may have.. forgotten to do that too. :cold_sweat: Apologies. :facepalm:
2. Yes, I agree. I don't think I'll fix it in the present state at the moment; I would like to use a better parsing library which can handle these kinds of situations more easily. Though, if I have some time maybe I'll take a look at it and see what I can do without too much effort.

**Reply by loofou** [reply]
I had pretty good experience with Pidgin, but it's a parser combinator as well, so in the end limited, especially if you try to parse more complex languages. I guess if you want to get serious, ANTLR is the way to go.

**Reply by exodrifter** [reply]
Experimental support for functions has been added in f1bdb4faabf495274d0166b0d28e61ef8b8a0a23.
